[Feature] Swap City buildings

danivenk opened the pull request:
This feature adds the ability choose which city specific city buildings belong to by clicking the highlight button in the city dialog and then clicking on the citybuilding.

To use this tool click on the cityhall of the city you want to add city buildings to.
Make sure you are using the public player.
Select the citybuildings of a different city that you want to add to the highlighted city, you can use clicking or dragging (it is a two click tool).

Highlight closes when the city dialog is closed or the highlight button is clicked on a different city dialog.

I am not sure if I implemented this the best way but I thought this was the best way to do it. Leartin has questioned why to only allow the pubic player to do this. Another question that could be asked is if this operation of changing which city a citybuilding belongs to should cost money or not.

If you have any suggestions/requirements or ideas of how it might need to change please feel free to say so.

danivenk committed:
add highlight of city_buildings on city_info click

danivenk committed:
Begin at change_city_of_citybuilding tool

danivenk committed:
The change city of city building now works.

danivenk committed:
Fixed selection after deselecting or dialogue dest

danivenk committed:
fixed bug by adding nullptr check

danivenk committed:
Merge branch 'OTRP-KUTAv6' of https://github.com/teamhimeh/simutrans into cbs-pull

danivenk commented:
TODO: unpress button of any open city dialogs which aren't the currently highlighted city.

I'm not quite sure how to alter the pressed state of buttons on other dialogs upon pressing a button, so I might need some help there.
![image](https://github.com/user-attachments/assets/2eb73cce-aa46-4e6f-9cc7-a0b7492b42e4)

teamhimeh commented on .gitignore:
```diff
@@ -28,7 +28,10 @@
 *.log
 /doxygen
-.vs/
+.*
+configure
+autom4te.cache
+simres.d.vs/
```
Please do not include the change which is not related to the main topic of this PR.

teamhimeh commented on simmenu.h:
```diff
@@ -82,6 +82,7 @@ enum {
 	TOOL_GENERATE_SCRIPT,
 	TOOL_REMOVE_HALT,
 	TOOL_EXTINGUISH_WAITING_GOODS,
+	TOOL_CHANGE_CITY_OF_CITYBUILDING,
```
(naming) `TOOL_CHANGE_CITY_OF_CITY_BUILDING` or `TOOL_CHANGE_CITY_OF_BUILDING`?

danivenk replied:
`TOOL_CHANGE_CITY_OF_CITY_BUILDING` is probably the best option since "City of Building" doesn't properly show what the tool does since there are also buildings that aren't linked to any city like the attractions etc.

teamhimeh commented on dataobj/environment.h:
```diff
@@ -313,6 +314,9 @@ class env_t
 	/// Hide buildings and trees within range of mouse cursor
 	static uint16 cursor_hide_range;
 
+	static bool highlight_city;
+	static stadt_t *highlighted_city;
```
Is `highlight_city` always false when `highlighted_city` is NULL? If so, only `highlighted_city` is enough.

teamhimeh replied:
Please reconsider the property name or add a description comment. It is unclear for what this property is.

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -371,7 +371,7 @@ void gebaeude_t::calc_image()
 image_id gebaeude_t::get_image() const
 {
-	if(env_t::hide_buildings!=0  &&  tile->has_image()) {
+	if(env_t::hide_buildings!=0 &&  tile->has_image()) {
```
Please follow the simutrans coding rule, although it is not strictly followed...
https://github.com/teamhimeh/simutrans/blob/6cb27bec30e16457af8b8db6797fa5690ab3e038/documentation/coding_styles.txt#L232-L237

```suggestion
	if(env_t::hide_buildings!=0  &&  tile->has_image()) {
```

teamhimeh commented on gui/city_info.cc:
```diff
@@ -197,6 +200,12 @@ void city_info_t::init()
 		add_component(&allow_growth);
+
+		// add "change highlight button"
+		highlight.init( button_t::box_state_automatic | button_t::flexible, "Highlight");
```
It is better to describe the purpose of this button.
```suggestion
		highlight.init( button_t::box_state_automatic | button_t::flexible, "Make building belong to");
```

danivenk replied:
The button itself only highlights the city buildings, at the same time it is selected the swap tool is selected which does swap/"make city building belong to" (altho the tool itself is only selected when the player is the public player).
See `city_info_t::action_triggered`.
![image](https://github.com/user-attachments/assets/aaaefacc-c669-4c97-a21d-b6b4c45b4af2)

teamhimeh replied:
In case of being a public player, highlighting the buildings is only the supplemental measure of selecting the city for the tool. The actual purpose is selecting the city for the tool, so the description should directly express it.
It may be helpful to alter the description text according to the selected player. When a not public player is selected, this button only highlights the buildings, so the description "Highlight" is appropriate for that case.

teamhimeh reviewed (commented):
Thank you for submitting the PR! Please address the comments.

teamhimeh commented:
> I'm not quite sure how to alter the pressed state of buttons on other dialogs upon pressing a button, so I might need some help there.

Probably checking `highlighted_city ` in `city_info_t::draw` is the best way. That function is called very frequently during the frame is shown.

danivenk commented:
> Probably checking highlighted_city  in city_info_t::draw is the best way. That function is called very frequently during the frame is shown.
Yeah I was thinking something like that too, I did also get a comment from prissi on the international discord:
![image](https://github.com/user-attachments/assets/4dc3df36-916f-43ca-8e10-6879eb487b23)
What is your opinion on that?

danivenk committed:
Match .gitignore to pull request base.

danivenk committed:
Fix coding style obj/gebaeude.cc

Co-authored-by: Himeshi <akimuhirai@gmail.com>

prissi commented:
This tool is a hack, as it cannot be used with scripts or online when relying on static variables.
You should have a deafult_param which identifies the city, probably by coordinate. This is transmitted to the tool by init.

On init, the tool asks for all city windows

// convert default_param to city
this_hightlight_city = welt->get_cities()[atoi(defualt_param)];
for( i=0: i<welt->get_cities()->get_count(); i++) {
  if(city_info_t *w=dynamic_cast<city_info_t *>(win_get_top((ptrdiff)welt->get_cities()[i]))) {
    w->update(this_hightlight_city);
  }
}

I would make this a general single click tool. Click on a house will transfer ownership to the current city in the default_param.

danivenk commented:
I guess I'll look into this later.
Altho what is your reasoning to make it a single click tool @prissi? I chose a double click one out of convenience to be able to just click and drag. Just like you can when building citybuildings or deleting in the OTRP version.
Having to click each single house individually is usually such a chore in my opinion.

teamhimeh commented:
Thank you @prissi for pointing it out. As prissi put it, the city to which the selected buildings belong has to be obtained via default_param to make the tool work on network games.
For the kind of the tool, I think two click tool is still appropriate because it is suitable to support the dragging selection of the buildings.

teamhimeh commented on simtool.cc:
```diff
@@ -7151,6 +7151,79 @@ const char *tool_stop_mover_t::do_work( player_t *player, const koord3d &last_po
+			}
+
+			stadt_t* old_city = gb->get_stadt();
+			stadt_t* new_city = env_t::highlighted_city;
```
https://github.com/teamhimeh/simutrans/pull/53#issuecomment-2362689120

This does not work in network games. Please use `default_param` instead.

danivenk committed:
changed tool name from TOOL_CHANGE_CITY_OF_CITYBUILDING to TOOL_CHANGE_CITY_OF_BUILDING

danivenk committed:
implemented default_param implementation of tool + fixing highlight without env_t

danivenk committed:
update button label based on active player

danivenk commented:
I have tried implementing the aforementioned changes. I'm not sure if they are all implemented correctly but please do tell me if that is the case.

danivenk committed:
forgotten to remove redundant env_t variables

danivenk committed:
forgot some more redundant code

danivenk committed:
added is_building_of_city method to also include monuments and HQs in tool

danivenk commented:
Added the monuments and HQ buildings to the ones that can be assigned to a new city as per request.

teamhimeh commented on simtool.h:
```diff
@@ -702,6 +705,29 @@ class tool_stop_mover_t : public two_click_tool_t {
 	void read_start_position(player_t *player, const koord3d &pos);
 };
 
+class tool_change_city_of_building_t : public two_click_tool_t {
```
Inheriting `two_click_kartenboden_tool_t` allows users to select only the ground tiles.
```suggestion
class tool_change_city_of_building_t : public two_click_kartenboden_tool_t {
```

teamhimeh commented on simtool.cc:
```diff
@@ -7150,6 +7150,92 @@ const char *tool_stop_mover_t::do_work( player_t *player, const koord3d &last_po
+bool tool_change_city_of_building_t::init(player_t* player) {
+	two_click_tool_t::init(player);
+
+	// convert default_param to city and update highlight_city
+	highlight_city = welt->get_cities()[atoi(default_param)];
```
Could you check whether the tool works with a network game?
`default_param` is usually set by the caller of the toll, using `set_default_param` function.

teamhimeh commented on gui/city_info.cc:
```diff
@@ -399,6 +423,32 @@ bool city_info_t::action_triggered( gui_action_creator_t *comp,value_t /* */)
+		// set display dirty and select tool
+		welt->set_dirty();
+		welt->set_background_dirty();
+		welt->set_tool( tool_t::general_tool[TOOL_CHANGE_CITY_OF_BUILDING], welt->get_public_player());
```
Probably `default_param` should be set here.

teamhimeh commented on simtool.cc:
```diff
@@ -7150,6 +7150,92 @@ const char *tool_stop_mover_t::do_work( player_t *player, const koord3d &last_po
+			w->update(highlight_city);
+		}
+	}
+	return true;
```
The function should return false here to reject the initialization if `highlight_city` is not valid.

teamhimeh commented on gui/city_info.cc:
```diff
@@ -399,6 +423,32 @@ bool city_info_t::action_triggered( gui_action_creator_t *comp,value_t /* */)
+
+		return true;
+	}
+	else if (	comp==&highlight && !highlight.pressed	) {
```
```suggestion
	else if (  comp==&highlight && !highlight.pressed  ) {
```

danivenk committed:
changed two click to two click kartenboden and fixed some formatting

danivenk committed:
set default param in city info dialog, return init false if highlight_city failed

danivenk committed:
Merge pull request #17 from teamhimeh/OTRP-KUTAv6

update cbs-pull with OTRP-KUTAv6

danivenk committed:
changed default_param from index of citylist to position of the city (inlcudes debugging printing will remove after getting it working with Network)

danivenk commented:
I can't seem to get it working in network mode, however the rest does seem to work so see this as a beta push since some debug printing is still there, expect a commit for removing that with the commit that fixes the network mode of the tool.

teamhimeh commented:
> I can't seem to get it working in network mode

It seems that the tool usage is correct. Is `default_param` incorrect in the network mode?

teamhimeh commented on gui/city_info.cc:
```diff
@@ -374,6 +389,12 @@ void city_info_t::update_labels()
 	lb_unemployed.buf().printf("%s: %d", translator::translate("Unemployed"), c->get_unemployed()); lb_unemployed.update();
 	lb_homeless.buf().printf("%s: %d", translator::translate("Homeless"), c->get_homeless());       lb_homeless.update();
+
+	if (welt->get_active_player_nr() == 1) {
```
(nit) Please use `PUBLIC_PLAYER_NR` instead of the magic number `1`.

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -373,7 +374,7 @@ image_id gebaeude_t::get_image() const
 {
 	if(env_t::hide_buildings!=0  &&  tile->has_image()) {
 		// opaque houses
-		if(is_city_building()) {
+		if(is_building_of_city()) {
```
This should not be changed. It affects the existing hiding feature.

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -416,22 +423,39 @@ FLAGGED_PIXVAL gebaeude_t::get_outline_colour() const
 {
 	uint8 colours[] = { COL_BLACK, COL_YELLOW, COL_YELLOW, COL_PURPLE, COL_RED, COL_GREEN };
 	FLAGGED_PIXVAL disp_colour = 0;
+	
+	// get city highlight & city highlighted
+	city_info_t* w_city = dynamic_cast<city_info_t*>(win_get_top());
+	bool city_highlight = w_city ? w_city->get_highlight() : 0;
+	stadt_t* city = w_city ? w_city->get_city() : nullptr;
+
 	if(env_t::hide_buildings!=env_t::NOT_HIDE) {
-		if(is_city_building()) {
+		if(is_building_of_city()) {
```
ditto

danivenk commented:
> > I can't seem to get it working in network mode
> 
> It seems that the tool usage is correct. Is `default_param` incorrect in the network mode?
![image](https://github.com/user-attachments/assets/ca08874f-9054-4c32-b691-f5d93527e50d)

In network mode, upon init of the tool the `default_param` is correct (the coords of the stadt) which makes the `highlight_city` variable of the tool correctly point to the city of the coords were set in the `default_param`. Even when then selecting an area of city buildings the `highlight_city` stays pointing to the correct city (afaik at least). However upon releasing the drag the `do_work` function never gets called and the `default_param` is already changed to the next value the game needs.
As seen in the code, the game should print to the console when `do_work` is called for this tool.

![image](https://github.com/user-attachments/assets/50219282-ed6a-4a13-90a4-c1736db3492d)
![image](https://github.com/user-attachments/assets/ee99ae31-d74e-479b-80a4-f3c360f0175f)
I have tried to see what happens when I put the debug on 2-3 and discovered some errors... Seems like something goes wrong with the network packets.

danivenk committed:
fixed accidental replacement of method

danivenk committed:
defined PUBLIC_PLAYER_NR and used instead of magic number

danivenk commented:
The nwc tool warning doesn't seem to be related.
But adding `bool is_work_network_safe() const OVERRIDE { return true; }` to the tool class makes the code work server side, however when trying to activate it client side it causes a disconnect/desync
Here are some of the console logs that might be of use
Server side:
![image](https://github.com/user-attachments/assets/5b8b30f8-4b64-4263-8321-36ec1d73ac41)
Client side:
![image](https://github.com/user-attachments/assets/c3176e8c-0bce-44a5-8d6a-1b99fdd74042)
![image](https://github.com/user-attachments/assets/14a08862-f1ac-4e59-86cc-5c4e7414c504)
![image](https://github.com/user-attachments/assets/b8d77e37-3649-4b69-a81c-b730872e7697)
Somewhere between the `do_work` and shortly after the client seems to desync...
The only idea I might come up with is the packets the clients sends to the server are missing some information. That is the only thing I can think of.

Anyways will give updates if anything comes up as usual.

danivenk committed:
removed debugging printing & fixed network mode execution crash

danivenk committed:
added work_on_ground function, made default_param parsing more robust and fixed tool selection bug when using hotkey to change player

danivenk committed:
removed redundant includes

danivenk commented:
AFAIK should the tool work in both single and multiplayer now and I have removed some bugs I had found. So it should be complete now as far as I am concerned.
If there are still things that are missing or should be changed in your opinion then please say so, especially if you notice any bugs.

teamhimeh commented on gui/city_info.cc:
```diff
@@ -23,6 +24,11 @@
 #define PAX_DEST_MIN_SIZE (16)      ///< minimum width/height of the minimap
 #define PAX_DEST_VERTICAL (4.0/3.0) ///< aspect factor where minimaps change to over/under instead of left/right
 
+#define PUBLIC_PLAYER_NR (1)
```
Please import `simplay.h` instead.

teamhimeh commented on simtool.cc:
```diff
@@ -7150,6 +7150,168 @@ const char *tool_stop_mover_t::do_work( player_t *player, const koord3d &last_po
+	if(  packet->is_loading()  ) {
+		packet->rdwr_short(p1);
+		packet->rdwr_short(p2);
+		if ( p1 >= 0 || p2 >= 0 ) {
```
[Q] Is the case like `(p1, p2) = (100, -100)` allowed?

danivenk replied:
oh right that is not supposed to be allowed, since Simutrans coordinates are always positive numbers afaik.

teamhimeh commented on simtool.h:
```diff
@@ -22,6 +22,9 @@
 
 #include "display/viewport.h"
 
+#include "gui/simwin.h"
+#include "gui/city_info.h"
```
Should these be in the `.cc` file rather than the header file?

danivenk replied:
noticed that those two were already in the `.cc` so I've removed them in the header file.

teamhimeh commented on simtool.cc:
```diff
@@ -7150,6 +7150,168 @@ const char *tool_stop_mover_t::do_work( player_t *player, const koord3d &last_po
+}
+
+
+void tool_change_city_of_building_t::rdwr_custom_data(memory_rw_t *packet)
```
I wonder whether we should really use `rdwr_custom_data`. Isn't `default_param` enough?

danivenk replied:
in its current form, the function seems to be necessary. I have tried running the tool without the custom data function, and as far as I remember it wouldn't work in Network mode when I removed the function.

However maybe some of the changes you suggested might fix that issue, so I'll investigate this.

danivenk replied:
Okay so after further investigation, the custom data & stadt_t pointer in the tool seems to be necessary for the desired working of the tool

teamhimeh commented on simtool.h:
```diff
@@ -702,6 +705,32 @@ class tool_stop_mover_t : public two_click_tool_t {
 
+class tool_change_city_of_building_t : public two_click_kartenboden_tool_t {
+private:
+	stadt_t* highlight_city = nullptr;
```
Since the highlighted city should be always restorable from `default_param`, I wonder we should really use the state property here.

danivenk replied:
Okay so after further investigation, the custom data & stadt_t pointer in the tool seems to be necessary for the desired working of the tool

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -403,7 +404,13 @@ image_id gebaeude_t::get_image() const
-	if(env_t::hide_buildings!=0  &&  env_t::hide_with_transparency  &&  !zeige_baugrube) {
+	// get city highlight & city highlighted
+	city_info_t* w_city = dynamic_cast<city_info_t*>(win_get_top());
+	bool city_highlight = w_city ? w_city->get_highlight() : 0;
```
(nit) prefer `false` than `0`.

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -403,7 +404,13 @@ image_id gebaeude_t::get_image() const
 {
-	if(env_t::hide_buildings!=0  &&  env_t::hide_with_transparency  &&  !zeige_baugrube) {
+	// get city highlight & city highlighted
+	city_info_t* w_city = dynamic_cast<city_info_t*>(win_get_top());
```
I think it's better to have the highlighted city as a static value of `env_t` rather than `dynamic_cast<city_info_t*>(win_get_top())` because the selected city is no longer highlighted once the city info window becomes not the top while still visible. Is it a deserved behavior?

danivenk replied:
It is something that I've kinda added due to the way prissi had suggested the implementation, not that should be the way. As long as the tool keeps working in network mode/when using scripts.
The question then becomes what if you have multiple city info dialogs open at the same time.
I guess that is fine when it just "unselects" all other cities when a different city is selected.

teamhimeh commented on gui/city_info.cc:
```diff
@@ -197,6 +203,16 @@ void city_info_t::init()
 		add_component(&allow_growth);
+
+		// add "change highlight button" based on active player
+		if (welt->get_active_player_nr() == 1) {
```
```suggestion
		if (welt->get_active_player()->is_public_service()) {
```

teamhimeh commented on gui/city_info.h:
```diff
@@ -77,6 +84,15 @@ class city_info_t : public gui_frame_t, private action_listener_t
+	stadt_t* get_city() { return city; };
+	
+	// is highlighted?
+	bool get_highlight() { return highlighted; };
```
Can't this be like this?
```suggestion
	bool is_highlighted() { return highlight.pressed; };
```

danivenk committed:
PUBLIC_PLAYER_NR fix and proper include

danivenk committed:
changed get_highlight to is_highlighted, fixed packet check, redundant include and bool 0 to false

teamhimeh commented on gui/city_info.h:
```diff
@@ -35,6 +36,8 @@ class city_info_t : public gui_frame_t, private action_listener_t
 	gui_textinput_t name_input;    ///< Input field where the name of the city can be changed
 	button_t allow_growth;         ///< Checkbox to enable/disable city growth
+	button_t highlight;			   ///< highlight button
+	bool highlighted = false;      ///< is currently highlighted
```
This property is no longer used in anywhere. Please remove it.

teamhimeh replied:
TODO: address this

danivenk committed:
changed the pointer for the highlighted city from the tool to a static env_t variable.
Also let highlight only work if the button is clicked

danivenk commented:
I have changed the code according to your feedback. As I've said in some resolved comments: after further investigation, the custom data & stadt_t pointer in the tool seems to be necessary for the desired working of the tool.

I have the static `env_t` variable now, not sure if it is entirely the right way but as far as I have tested it it seems to work fine now in network mode as well. I have not yet tested it as a script (which any tool should be), but for that I'd either need to figure out how to write a script or you can just test that for me if you feel like it. Whatever suits you.

EDIT: I don't think that saving the static variable in the savefile cause technically there is no guarantee that a pointer points to the right part upon restart.

teamhimeh commented on gui/city_info.cc:
```diff
@@ -383,6 +406,9 @@ void city_info_t::draw(scr_coord pos, scr_size size)
 	gui_frame_t::draw(pos, size);
+
+	// update pressed if hightlighted city is this->city
+	highlight.pressed = (env_t::highlighted_city && (env_t::highlighted_city == city)) ? true : false;
```
```suggestion
	highlight.pressed = env_t::highlighted_city  &&  (env_t::highlighted_city == city);
```

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -403,7 +404,13 @@ image_id gebaeude_t::get_image() const
 {
-	if(env_t::hide_buildings!=0  &&  env_t::hide_with_transparency  &&  !zeige_baugrube) {
+	// get city highlight & city highlighted
+	city_info_t *w_city = dynamic_cast<city_info_t*>(win_get_magic((ptrdiff_t)this->get_stadt()));
+	bool city_highlight = w_city ? w_city->is_highlighted() : false;
+	stadt_t* city = w_city ? w_city->get_city() : nullptr;
```
We can simply check this with `env_t::highlighted_city` instead of checking the state of the city info window.

teamhimeh commented:
> I don't think that saving the static variable in the savefile cause technically there is no guarantee that a pointer points to the right part upon restart.

Although it is optional, I think it's better to store and restore the highlighted city considering the fact that the network game is frequently saved and loaded. One of the ways to persist the highlighted city is to use the name of the city.

danivenk committed:
changed highlighting based on the highlighted_city env_t variable

danivenk commented:
I have added the changes you requested except for the saving/storing.

```
	if (  file->get_OTRP_version()>=33  ) {
		sint16 p1(-100), p2(-100);
		karte_t* welt = world();
		std::cout << welt << std::endl;
		if(  file->is_loading()  ) {
			file->rdwr_short(p1);
			file->rdwr_short(p2);
			if ( p1 >= 0 && p2 >= 0 ) {
				koord city_koord(p1, p2);

				for( int i = 0; i < welt->get_cities().get_count(); i++) {
					std::cout << highlighted_city->get_name() << " - " << welt->get_cities()[i]->get_name() << std::endl;
					if ( city_koord == welt->get_cities()[i]->get_pos() ) {
						highlighted_city = welt->get_cities()[i];
					}
				}
			}
		} else {
			p1 = highlighted_city->get_pos().x;
			p2 = highlighted_city->get_pos().y;
			if ( p1 >= 0 && p2 >= 0 ) {
				file->rdwr_short(p1);
				file->rdwr_short(p2);
			}
		}
	}
```
Above is the code I have been trying since this is also te way the custom data is sent and I thought it should work just fine, however there is one problem. It depends on the welt variable being properly assigned which doesn't exist as far as `environment.cc` is concerned. As you can see I have tried to do it as I have spotted in other places but this doesn't seem to work correctly. This almost seemed to work except for the fact that the "saved" highlighted_city seemed to change each time you reload the savegame.

teamhimeh commented on dataobj/environment.h:
```diff
@@ -313,6 +314,8 @@ class env_t
 	/// Hide buildings and trees within range of mouse cursor
 	static uint16 cursor_hide_range;
 
+	static stadt_t* highlighted_city;
```
Please add the description comment.

teamhimeh committed:
Use default_param to handle the city

teamhimeh commented:
I tested the current implmentation by myself and found that the beloging city was not changed correctly on the network game. I fixed by myself and the fix is in the link below.
https://github.com/teamhimeh/simutrans/commit/7ccf0b58f3610c3638f0f95bf2f897c16ee94f1f
With this implementation, I used only `default_param` to handle the highlighted city in `tool_change_city_of_building_t`. The implementation of some functions are greatly simplified. Please check it.

teamhimeh commented:
> It depends on the welt variable being properly assigned which doesn't exist as far as environment.cc is concerned.

In that case, it may be better to have the highlighted city as a `koord` object in `env_t`.

*FYI:* `koord` has `rdwr` function, `void rdwr(loadsave_t *file);`. You can easily persist a koord object with this.

danivenk commented:
> I tested the current implmentation by myself and found that the beloging city was not changed correctly on the network game. I fixed by myself and the fix is in the link below. [7ccf0b5](https://github.com/teamhimeh/simutrans/commit/7ccf0b58f3610c3638f0f95bf2f897c16ee94f1f) With this implementation, I used only `default_param` to handle the highlighted city in `tool_change_city_of_building_t`. The implementation of some functions are greatly simplified. Please check it.

Do you have an idea what the problem exactly was? And I take it I should then just cherry pick this commit to the branch for this pull request right?

danivenk commented:
> > It depends on the welt variable being properly assigned which doesn't exist as far as environment.cc is concerned.
> 
> In that case, it may be better to have the highlighted city as a `koord` object in `env_t`.
> 
> _FYI:_ `koord` has `rdwr` function, `void rdwr(loadsave_t *file);`. You can easily persist a koord object with this.

Okay I will do that when I have the time.

teamhimeh commented:
> Do you have an idea what the problem exactly was?

Sorry, I don't have any idea for it. `rdwr_custom_data` should have restored the position of the highlighted city, but it didn't.
Anyway, please just cherry-pick it then my fix is applied. Please check if it works fine by yourself just in case.

danivenk commented:
> > Do you have an idea what the problem exactly was?
> 
> Sorry, I don't have any idea for it. `rdwr_custom_data` should have restored the position of the highlighted city, but it didn't. Anyway, please just cherry-pick it then my fix is applied. Please check if it works fine by yourself just in case.

That is indeed weird, I will cherrypick and test it! Thank you for your work.

danivenk committed:
added comment to env_t header entry

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -45,6 +45,7 @@ static pthread_mutex_t add_to_city_mutex = PTHREAD_MUTEX_INITIALIZER;
 #include "../dataobj/environment.h"
 
 #include "../gui/obj_info.h"
+#include "../gui/city_info.h"
```
Probably this `include` is no longer needed.

teamhimeh commented:
Please address https://github.com/teamhimeh/simutrans/pull/53#discussion_r1896352323

danivenk committed:
removed redudant code

danivenk committed:
changed highlighted_city from stadt_t* to koord

danivenk commented:
K I have changed the things you have asked but got stuck at changing the env_t variable to a koord. For me [76e3e0a](https://github.com/teamhimeh/simutrans/pull/53/commits/76e3e0ab74c461572955bf921b95b38e1b0f9dd3) works fine but when I compile & run [d6cfcdd](https://github.com/teamhimeh/simutrans/pull/53/commits/d6cfcdde9ec935c8a920e00550d29ec45a43ffb1) the game crashed before a game is loaded. When I run debug to level 5 still nothing shows up for me. Could you please confirm that [d6cfcdd](https://github.com/teamhimeh/simutrans/pull/53/commits/d6cfcdde9ec935c8a920e00550d29ec45a43ffb1) also doesn't work for you? If it does then I don't have a clue why it suddenly doesn't work anymore when I basically only changed the value/type of the variable (and all references/comparisons of course appropriately)

In summary, in my case:
[76e3e0a](https://github.com/teamhimeh/simutrans/pull/53/commits/76e3e0ab74c461572955bf921b95b38e1b0f9dd3) - works fine
[d6cfcdd](https://github.com/teamhimeh/simutrans/pull/53/commits/d6cfcdde9ec935c8a920e00550d29ec45a43ffb1) - crashes before loading save and after loading pakset


Hence why this doesn't contain the save to file part, altho that shouldn't be too hard as you have already mentioned. Do you already have an idea for the version number I can use?. So we don't need to change that afterwards again (hopefully).

teamhimeh commented:
Generally speaking, the cause of the crash is determined with the stack trace issued by the debugger (e.g. gdb). Did you get the stack trace of the crash?

teamhimeh commented:
I got the stack trace. Here is the part of it.

```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7c)
  * frame #0: 0x00000001001a62e4 sim`gebaeude_t::get_outline_image() const [inlined] stadt_t::get_pos(this=0x0000000000000000) const at simcity.h:558:39 [opt]
    frame #1: 0x00000001001a62e4 sim`gebaeude_t::get_outline_image(this=0x00006000029ffa50) const at gebaeude.cc:406:92 [opt]
    frame #2: 0x00000001001b3110 sim`obj_t::display(this=0x00006000029ffa50, xpos=<unavailable>, ypos=<unavailable>, clip_num='\a') const at simobj.cc:198:33 [opt]
    frame #3: 0x0000000100052060 sim`objlist_t::display_obj_bg(short, short, unsigned char, signed char) const [inlined] local_display_obj_bg(obj=<unavailable>, xpos=<unavailable>, ypos=<unavailable>, clip_num='\a') at objlist.cc:1177:8 [opt]
    frame #4: 0x0000000100052048 sim`objlist_t::display_obj_bg(this=0x00000001580755d0, xpos=640, ypos=-64, start_offset=<unavailable>, clip_num='\a') const at objlist.cc:1193:9 [opt]
    frame #5: 0x0000000100039f88 sim`grund_t::display_obj_all(this=0x00000001580755c8, xpos=640, ypos=-64, raster_tile_width=128, is_global=true, clip_num=<unavailable>) const at grund.cc:1482:26 [opt]
    frame #6: 0x00000001002f11f0 sim`planquadrat_t::display_obj(this=0x0000000160b5e3a8, xpos=640, ypos=-64, raster_tile_width=128, is_global=true, hmin='\x02', hmax='\x7f', clip_num='\a') const at simplan.cc:492:9 [opt]
    frame #7: 0x0000000100091f44 sim`main_view_t::display_region(this=<unavailable>, lt=<unavailable>, wh=<unavailable>, y_min=<unavailable>, y_max=<unavailable>, (null)=<unavailable>, threaded=true, clip_num='\a') at simview.cc:0 [opt]
    frame #8: 0x0000000100092bf0 sim`main_view_t::display(this=0x00006000024f7080, force_dirty=<unavailable>) at simview.cc:264:3 [opt]
```

The cause of this issue is that `get_stadt()` may return NULL at `gebaeude.cc:406`. I think doing a NULL check there fixes the issue.

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -403,7 +403,8 @@ image_id gebaeude_t::get_image() const
 image_id gebaeude_t::get_outline_image() const
 {
-	if(env_t::hide_buildings!=0  &&  env_t::hide_with_transparency  &&  !zeige_baugrube) {
+	if (((env_t::hide_buildings != 0 && env_t::hide_with_transparency) || (this->get_stadt()->get_pos() == env_t::highlighted_city)) && !zeige_baugrube)
```
The null check of `get_stadt()` is needed.

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -425,13 +427,19 @@ FLAGGED_PIXVAL gebaeude_t::get_outline_colour() const
 image_id gebaeude_t::get_image(int nr) const
 {
-	if(zeige_baugrube || env_t::hide_buildings) {
+	if(zeige_baugrube || env_t::hide_buildings || ((this->get_stadt()->get_pos() == env_t::highlighted_city) && (env_t::highlighted_city != koord::invalid))) {
```
ditto

danivenk commented:
ah of course, completely forgot that before it was fine without a check because if city was null highlighted_city would become just null. However now we call a method of city which needs to not be null. Stupid mistake on my end.

I will try to get the save part working today and then it should be basically done right? Or am I missing things?

danivenk committed:
added missing null check for the get_stadt() method usage.

teamhimeh commented:
> I will try to get the save part working today and then it should be basically done right? Or am I missing things?

Sorry. Although I suggested to put `highlighted_city` in `env_t`, I found that `env_t` is the place for the persisted values which are independent from the save data. Since `highlighted_city` is dependent on the specific map, `env_t` is not a suitable place. If we persist the coordinate of `highlighted_city` with `env_t`, the coordinate may be restored for a different map, which may result in an unexpected behavior.

I think persisting `is_highlighted` with `city_info_t::rdwr(loadsave_t*)` is a good way. `highlighted_city` can be just a static variable of `city_info_t`. In the loading, if `is_highlighted` is loaded as true, `city_info_t::highlighted_city` should be that city and the city changing tool is to be set.

danivenk commented:
Oh and bit of a shame I had just changed it, I will do it like that then. Do we also convert the variable in env_t back to a `stadt_t` pointer then?

teamhimeh commented:
Yes, there is no reason to keep `highlighted_city` as a `koord`.

danivenk committed:
Revert "added missing null check for the get_stadt() method usage."

This reverts commit 1db72e75399752a70dcb4a0f9357750ff65b552b.

danivenk committed:
Revert "changed highlighted_city from stadt_t* to koord"

This reverts commit d6cfcdde9ec935c8a920e00550d29ec45a43ffb1.

danivenk committed:
moved the highlighted city from env_t to city_info_t

danivenk commented:
I have converted I have moved the highlighted_city from the `env_t` class to the `city_info_t` class. However whenever I try to add any state related to the variable to the `city_info_t::rdwr` the game saves complete, but when I load them I get a "not compatible" error. When I try to investigate it a bit more I notice that the save game version is somehow converted to the max value of a `uint32`, which value is used in Simutrans for `INVALID_FILE_VERSION` (it is defined as follows: `#define INVALID_FILE_VERSION 0xFFFFFFFFu`).
Tbh I don't have a clue why this happens, but it seems like it happens when I for example call `is_highlighted` in `city_info_t::rdwr` or set `highlighted_city` in `city_info_t::rdwr`.
You'd think it shouldn't be a problem to do those things inside the `city_info_t::rdwr` function.

teamhimeh commented:
As far as I know, `rdwr` functions for the gui related classes (inheriting `gui_frame_t`) does not handle the version compatibility, because they are for restoring the gui only in the network mode. Version checking is not needed in `city_info_t::rdwr`, and please check whether if there is nothing wrong in the save data version compatibility.

danivenk commented:
I know that, I think you misunderstood what I meant.
With the way the code is now (on my side) when I save the game it seems to work fine, when I then try to load a different save game from the load menu the incompatible error shows up. When I did some investigation I found that the when read the savegame loads somehow were unable to load because the file version basically has become the equivalent of null/-1, ie something has gone wrong. The only code I have changed for it to go wrong is the code related to saving the highlighted state like you requested. Hence my mention of the function in which that is done.

I think the problem might be indeed a data issue, which I might need to do a check for, but I need to investigate first how I can do that (or even what happens when a rdwr to file goes wrong). I'll report back with any update or if I can't find a fix again.

danivenk committed:
tried adding highlight button state to city_info_t::rdwr

danivenk commented:
okay so it was a different issue, I have been able to seemingly add it to the rdwr function, however the function seems only to be called upon saving. When loading a game the data doesn't seem to be loaded. I'm not quite sure why that would be the case, or am I missing something?

danivenk committed:
added some checks and init a dialog on loading

danivenk committed:
no need to create a dialog in this if statement

danivenk commented:
The window recreation after load works (if `env_t::restore_UI` is true), however, I don't have the reselect of the tool working just yet. Not sure why it won't select the tool upon loading. I have done some debug prints and the default param part is set fine, just the tool doesn't seem to get set in the `city_info_t::rdwr` function.

teamhimeh commented on dataobj/environment.h:
```diff
@@ -11,6 +11,7 @@
 #include "../simtypes.h"
 #include "../simconst.h"
 #include "../simcolor.h"
+#include "../simcity.h"
```
TODO: remove this

teamhimeh commented on gui/city_info.cc:
```diff
@@ -23,6 +25,9 @@
 #define PAX_DEST_VERTICAL (4.0/3.0) ///< aspect factor where minimaps change to over/under instead of left/right
 
+tool_change_city_of_building_t* city_info_t::citybuilding_tool=new tool_change_city_of_building_t();
+cbuffer_t city_info_t::param_str;
```
TODO: enclose param_str in tool_change_city_of_building_t

teamhimeh commented on gui/city_info.cc:
```diff
@@ -197,6 +202,16 @@ void city_info_t::init()
 		add_component(&allow_growth);
+
+		// add "change highlight button" based on active player
+		if (welt->get_active_player_nr() == PUBLIC_PLAYER_NR) {
```
TODO: use is_public_service()

teamhimeh commented on gui/city_info.cc:
```diff
@@ -374,6 +392,12 @@ void city_info_t::update_labels()
 	lb_unemployed.buf().printf("%s: %d", translator::translate("Unemployed"), c->get_unemployed()); lb_unemployed.update();
 	lb_homeless.buf().printf("%s: %d", translator::translate("Homeless"), c->get_homeless());       lb_homeless.update();
+
+	if (welt->get_active_player_nr() == PUBLIC_PLAYER_NR) {
```
TODO: ditto

teamhimeh commented on gui/city_info.cc:
```diff
@@ -457,6 +512,26 @@ void city_info_t::rdwr(loadsave_t *file)
 
 	year_month_tabs.rdwr(file);
 
+	bool highlighted = false;
+	if (file->is_saving()) {
+		highlighted = is_highlighted();
+	}
```
TODO: 
```suggestion
	bool highlighted = is_highlighted();
```

teamhimeh commented on gui/city_info.cc:
```diff
@@ -457,6 +512,26 @@ void city_info_t::rdwr(loadsave_t *file)
+	file->rdwr_bool( highlighted );
+
+	highlight.pressed = highlighted;
+	highlighted_city = highlighted ? city : nullptr;
```
TODO: probably this should be
```c++
if(highlighted) { highlighted_city = city; }
```

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -403,7 +403,13 @@ image_id gebaeude_t::get_image() const
 {
-	if(env_t::hide_buildings!=0  &&  env_t::hide_with_transparency  &&  !zeige_baugrube) {
+	// get city highlight & city highlighted
+	city_info_t *w_city = dynamic_cast<city_info_t*>(win_get_magic((ptrdiff_t)this->get_stadt()));
+	bool city_highlight = w_city ? w_city->is_highlighted() : false;
+	stadt_t* city = city_highlight ? w_city->get_city() : nullptr;
```
TODO: add `const` and merge conditions

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -416,6 +422,12 @@ FLAGGED_PIXVAL gebaeude_t::get_outline_colour() const
 	FLAGGED_PIXVAL disp_colour = 0;
+
+	// get city highlight & city highlighted
+	city_info_t *w_city = dynamic_cast<city_info_t*>(win_get_magic((ptrdiff_t)this->get_stadt()));
+	bool city_highlight = w_city ? w_city->is_highlighted() : false;
+	stadt_t* city = city_highlight ? w_city->get_city() : nullptr;
```
TODO: ditto

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -425,13 +437,24 @@ FLAGGED_PIXVAL gebaeude_t::get_outline_colour() const
 {
-	if(zeige_baugrube || env_t::hide_buildings) {
+	// get city highlight & city highlighted
+	city_info_t *w_city = dynamic_cast<city_info_t*>(win_get_magic((ptrdiff_t)this->get_stadt()));
+	bool city_highlight = w_city ? w_city->is_highlighted() : false;
+	stadt_t* city = city_highlight ? w_city->get_city() : nullptr;
+
```
TODO: ditto

teamhimeh commented on obj/gebaeude.cc:
```diff
@@ -445,7 +468,14 @@ image_id gebaeude_t::get_front_image() const
-	if (env_t::hide_buildings != 0   &&  (is_city_building()  ||  (env_t::hide_buildings == env_t::ALL_HIDDEN_BUILDING  &&  tile->get_desc()->get_type() < building_desc_t::others))) {
+	
+	// get city highlight & city highlighted
+	city_info_t *w_city = dynamic_cast<city_info_t*>(win_get_magic((ptrdiff_t)this->get_stadt()));
+	bool city_highlight = w_city ? w_city->is_highlighted() : false;
```
TODO: ditto

teamhimeh commented on gui/city_info.cc:
```diff
@@ -457,6 +512,26 @@ void city_info_t::rdwr(loadsave_t *file)
+		// set display dirty and select tool
+		welt->set_dirty();
+		welt->set_background_dirty();
+		welt->set_tool( citybuilding_tool, welt->get_public_player());
```
TODO: Investigate why it does not work

teamhimeh replied:
When reloading the game, all tools are deselected. This tool should not be selected as well.

teamhimeh commented:
> The window recreation after load works (if env_t::restore_UI is true), however, I don't have the reselect of the tool working just yet. Not sure why it won't select the tool upon loading.

I don't know why it does not work in `rdwr` either, but I'll investigate by myself since it is a minor issue.
Since there is no major issue left, I merge this PR. Thank you for your contribution!
